<a href="https://colab.research.google.com/github/akbarymam/Awesome-Projects/blob/main/Analisa_Data_KEL_12.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### KELOMPOK 12
1. Aflah Iram Athena - 1305213035
2. Imam Akbar Nugraha - 1305213056


# **LIbrary**

In [ ]:
import pandas as pd
import numpy as np
import scipy.stats as stats
from  numpy.linalg import inv
from scipy.linalg import solve

#**Data**

Membaca Data

In [ ]:
dataset = pd.read_csv("/content/Converted.csv", delimiter=";", skipinitialspace=True)
dataset

Country        Area  Birth rate Current account balance   \
0             Japan     377,835         947      170,200,000,000.00    
1           Germany     357,021         833       73,590,000,000.00    
2      Saudi Arabia   1,960,582        2956       51,500,000,000.00    
3            Russia  17,075,200          98       46,040,000,000.00    
4       Switzerland       41,29         977       40,950,000,000.00    
..              ...         ...         ...                      ...   
144           Italy      301,23         889      (21,100,000,000.00)   
145           Spain     504,782         101      (30,890,000,000.00)   
146  United Kingdom      244,82        1078      (33,460,000,000.00)   
147       Australia   7,686,850        1226      (38,300,000,000.00)   
148   United States   9,631,418        1414     (646,500,000,000.00)   

     Death rate Electricity consumption Electricity production   \
0           895         971,000,000,000   1,044,000,000,000.00    
1          1055         519,500,000,000     560,000,000,000.00    
2           262         128,500,000,000     138,200,000,000.00    
3          1452         894,300,000,000     915,000,000,000.00    
4           848          54,530,000,000      63,470,000,000.00    
..          ...                     ...                     ...   
144         103         293,900,000,000     261,600,000,000.00    
145         963         218,400,000,000     229,000,000,000.00    
146        1018         337,400,000,000     395,900,000,000.00    
147         744         195,600,000,000     210,300,000,000.00    
148         825       3,660,000,000,000   3,839,000,000,000.00    

                 Exports                     GDP GDP per capita   ...  \
0    538,800,000,000.00    3,745,000,000,000.00       29,400.00   ...   
1    893,300,000,000.00    2,362,000,000,000.00       28,700.00   ...   
2    113,000,000,000.00      310,200,000,000.00       12,000.00   ...   
3    162,500,000,000.00    1,408,000,000,000.00        9,800.00   ...   
4    130,700,000,000.00      251,900,000,000.00       33,800.00   ...   
..                   ...                     ...             ...  ...   
144  336,400,000,000.00    1,609,000,000,000.00       27,700.00   ...   
145  172,500,000,000.00      937,600,000,000.00       23,300.00   ...   
146  347,200,000,000.00    1,782,000,000,000.00       29,600.00   ...   
147   86,890,000,000.00      611,700,000,000.00       30,700.00   ...   
148  795,000,000,000.00   11,750,000,000,000.00       40,100.00   ...   

     Life expectancy at birth Military expenditures Natural gas consumption   \
0                        8115                 100.0       80,420,000,000.00    
1                        7865                 150.0       99,550,000,000.00    
2                        7546                1000.0       56,400,000,000.00    
3                         671                   NaN      405,800,000,000.00    
4                        8039                 100.0        3,093,000,000.00    
..                        ...                   ...                      ...   
144                      7968                 180.0       71,180,000,000.00    
145                      7952                 120.0       17,960,000,000.00    
146                      7838                 240.0       92,850,000,000.00    
147                      8039                 270.0       23,330,000,000.00    
148                      7771                 330.0      640,900,000,000.00    

     Oil consumption    Population   Public debt  Railways  \
0          5,290,000   127,417,244        1643.0   23705.0   
1          2,891,000    82,431,390         658.0   46039.0   
2          1,550,000    26,417,599          75.0    1392.0   
3          2,310,000   143,420,309         282.0   87157.0   
4               290,4    7,489,370         572.0    4533.0   
..                ...           ...          ...       ...   
144        1,866,000    58,103,033        1056.0   19507.0   
145     

Memilih Kolom yang akan dianalisis

In [ ]:
data_dipake = ["GDP", "Exports", "Imports", "Industrial production growth rate", "Investment", "Unemployment rate"]
df = dataset[data_dipake]
df

GDP              Exports                Imports  \
0     3,745,000,000,000.00   538,800,000,000.00     401,800,000,000.00    
1     2,362,000,000,000.00   893,300,000,000.00     716,700,000,000.00    
2       310,200,000,000.00   113,000,000,000.00      36,210,000,000.00    
3     1,408,000,000,000.00   162,500,000,000.00      92,910,000,000.00    
4       251,900,000,000.00   130,700,000,000.00     121,100,000,000.00    
..                      ...                  ...                    ...   
144   1,609,000,000,000.00   336,400,000,000.00     329,300,000,000.00    
145     937,600,000,000.00   172,500,000,000.00     222,000,000,000.00    
146   1,782,000,000,000.00   347,200,000,000.00     439,400,000,000.00    
147     611,700,000,000.00    86,890,000,000.00      98,100,000,000.00    
148  11,750,000,000,000.00   795,000,000,000.00   1,476,000,000,000.00    

     Industrial production growth rate  Investment  Unemployment rate  
0                                 66.0        24.0               47.0  
1                                 22.0       176.0              106.0  
2                                 28.0       172.0               25.0  
3                                 64.0       191.0               83.0  
4                                 47.0       204.0               34.0  
..                                 ...         ...                ...  
144                                7.0       193.0               86.0  
145                                3.0       254.0              104.0  
146                                9.0       162.0               48.0  
147                               19.0       253.0               51.0  
148                               44.0       157.0               55.0  

[149 rows x 6 columns]

Mengganti Nama Kolom dengan variabel regresi

In [ ]:
df = df.rename(columns = {'GDP':'Y','Exports':'X1','Imports':'X2','Industrial production growth rate':'X3','Investment':'X4','Unemployment rate':'X5'})
df

Y                   X1                     X2    X3  \
0     3,745,000,000,000.00   538,800,000,000.00     401,800,000,000.00   66.0   
1     2,362,000,000,000.00   893,300,000,000.00     716,700,000,000.00   22.0   
2       310,200,000,000.00   113,000,000,000.00      36,210,000,000.00   28.0   
3     1,408,000,000,000.00   162,500,000,000.00      92,910,000,000.00   64.0   
4       251,900,000,000.00   130,700,000,000.00     121,100,000,000.00   47.0   
..                      ...                  ...                    ...   ...   
144   1,609,000,000,000.00   336,400,000,000.00     329,300,000,000.00    7.0   
145     937,600,000,000.00   172,500,000,000.00     222,000,000,000.00    3.0   
146   1,782,000,000,000.00   347,200,000,000.00     439,400,000,000.00    9.0   
147     611,700,000,000.00    86,890,000,000.00      98,100,000,000.00   19.0   
148  11,750,000,000,000.00   795,000,000,000.00   1,476,000,000,000.00   44.0   

        X4     X5  
0     24.0   47.0  
1    176.0  106.0  
2    172.0   25.0  
3    191.0   83.0  
4    204.0   34.0  
..     ...    ...  
144  193.0   86.0  
145  254.0  104.0  
146  162.0   48.0  
147  253.0   51.0  
148  157.0   55.0  

[149 rows x 6 columns]

Konversi Kolom menjadi numerik

In [ ]:
for col in df.columns:
    if df[col].dtype == "object":
        df[col] = df[col].str.replace(",", "").astype(float)

df

Y            X1            X2    X3     X4     X5
0    3.745000e+12  5.388000e+11  4.018000e+11  66.0   24.0   47.0
1    2.362000e+12  8.933000e+11  7.167000e+11  22.0  176.0  106.0
2    3.102000e+11  1.130000e+11  3.621000e+10  28.0  172.0   25.0
3    1.408000e+12  1.625000e+11  9.291000e+10  64.0  191.0   83.0
4    2.519000e+11  1.307000e+11  1.211000e+11  47.0  204.0   34.0
..            ...           ...           ...   ...    ...    ...
144  1.609000e+12  3.364000e+11  3.293000e+11   7.0  193.0   86.0
145  9.376000e+11  1.725000e+11  2.220000e+11   3.0  254.0  104.0
146  1.782000e+12  3.472000e+11  4.394000e+11   9.0  162.0   48.0
147  6.117000e+11  8.689000e+10  9.810000e+10  19.0  253.0   51.0
148  1.175000e+13  7.950000e+11  1.476000e+12  44.0  157.0   55.0

[149 rows x 6 columns]

Menghapus missing Value dalam dataframe

In [ ]:
df = df.dropna()
df.reset_index(drop=True, inplace=True)
df

Y            X1            X2    X3     X4     X5
0    3.745000e+12  5.388000e+11  4.018000e+11  66.0   24.0   47.0
1    2.362000e+12  8.933000e+11  7.167000e+11  22.0  176.0  106.0
2    3.102000e+11  1.130000e+11  3.621000e+10  28.0  172.0   25.0
3    1.408000e+12  1.625000e+11  9.291000e+10  64.0  191.0   83.0
4    2.519000e+11  1.307000e+11  1.211000e+11  47.0  204.0   34.0
..            ...           ...           ...   ...    ...    ...
111  1.609000e+12  3.364000e+11  3.293000e+11   7.0  193.0   86.0
112  9.376000e+11  1.725000e+11  2.220000e+11   3.0  254.0  104.0
113  1.782000e+12  3.472000e+11  4.394000e+11   9.0  162.0   48.0
114  6.117000e+11  8.689000e+10  9.810000e+10  19.0  253.0   51.0
115  1.175000e+13  7.950000e+11  1.476000e+12  44.0  157.0   55.0

[116 rows x 6 columns]

Re-check missing value dari setiap kolom

In [ ]:
print(df.isnull().sum())

Y     0
X1    0
X2    0
X3    0
X4    0
X5    0
dtype: int64


# **Model Regresi Linear Berganda**

inisialisasi Kolom menjadi list

In [ ]:
list_Y = df["Y"]
list_X1 = df["X1"]
list_X2 = df["X2"]
list_X3 = df["X3"]
list_X4 = df["X4"]
list_X5 = df["X5"]

## Perhitungan Sigma Variabel regresi

In [ ]:
jum_Y = 0
for i in range (len(list_Y)):
  temp = list_Y[i]
  jum_Y += temp
print(jum_Y)

54384142000000.0


In [ ]:
jum_X1 = 0
for i in range (len(list_X1)):
  temp = list_X1[i]
  jum_X1 += temp

print(jum_X1)

8618245200000.0


In [ ]:
jum_X2 = 0
for i in range (len(list_X2)):
  temp = list_X2[i]
  jum_X2 += temp

print(jum_X2)

8617518000000.0


In [ ]:
jum_X3 = 0
for i in range (len(list_X3)):
  temp = list_X3[i]
  jum_X3 += temp

print(jum_X3)

4560.0


In [ ]:
jum_X4 = 0
for i in range (len(list_X4)):
  temp = list_X4[i]
  jum_X4 += temp

print(jum_X4)

21551.0


In [ ]:
jum_X5 = 0
for i in range (len(list_X5)):
  temp = list_X5[i]
  jum_X5 += temp

print(jum_X5)

8322.0


In [ ]:
jum_Ypangkat2 = 0

for i in range (len(list_Y)):
  temp = list_Y[i]**2
  jum_Ypangkat2 += temp

print(jum_Ypangkat2)

2.430410745013239e+26


In [ ]:
jum_X1pangkat2 = 0

for i in range (len(list_X1)):
  temp = list_X1[i]**2
  jum_X1pangkat2 += temp

print(jum_X1pangkat2)

3.175863392746079e+24


In [ ]:
jum_X2pangkat2 = 0

for i in range (len(list_X2)):
  temp = list_X2[i]**2
  jum_X2pangkat2 += temp

print(jum_X2pangkat2)

4.221715336144281e+24


In [ ]:
jum_X3pangkat2 = 0

for i in range (len(list_X3)):
  temp = list_X3[i]**2
  jum_X3pangkat2 += temp

print(jum_X3pangkat2)

379042.0


In [ ]:
jum_X4pangkat2 = 0

for i in range (len(list_X4)):
  temp = list_X4[i]**2
  jum_X4pangkat2 += temp

print(jum_X4pangkat2)

5079731.0


In [ ]:
jum_X5pangkat2 = 0

for i in range (len(list_X5)):
  temp = list_X5[i]**2
  jum_X5pangkat2 += temp

print(jum_X5pangkat2)

1067234.0


In [ ]:
jum_X1xX2 = 0

for i in range (len(list_X1)):
  temp = list_X1[i] * list_X2[i]
  jum_X1xX2 += temp

print(jum_X1xX2)

3.4211320093785e+24


In [ ]:
jum_X1xX3 = 0

for i in range (len(list_X1)):
  temp = list_X1[i] * list_X3[i]
  jum_X1xX3 += temp

print(jum_X1xX3)

412566554100000.0


In [ ]:
jum_X1xX4 = 0

for i in range (len(list_X1)):
  temp = list_X1[i] * list_X4[i]
  jum_X1xX4 += temp

print(jum_X1xX4)

1455976732000000.0


In [ ]:
jum_X1xX5 = 0

for i in range (len(list_X1)):
  temp = list_X1[i] * list_X5[i]
  jum_X1xX5 += temp

print(jum_X1xX5)

570587119100000.0


In [ ]:
jum_X2xX3 = 0

for i in range (len(list_X1)):
  temp = list_X2[i] * list_X3[i]
  jum_X2xX3 += temp

print(jum_X2xX3)

407058437500000.0


In [ ]:
jum_X2xX4 = 0

for i in range (len(list_X2)):
  temp = list_X2[i] * list_X4[i]
  jum_X2xX4 += temp

print(jum_X2xX4)

1455694163600000.0


In [ ]:
jum_X2xX5 = 0

for i in range (len(list_X2)):
  temp = list_X2[i] * list_X5[i]
  jum_X2xX5 += temp

print(jum_X2xX5)

565991589700000.0


In [ ]:
jum_X3xX4 = 0

for i in range (len(list_X3)):
  temp = list_X3[i] * list_X4[i]
  jum_X3xX4 += temp

print(jum_X3xX4)

856478.0


In [ ]:
jum_X3xX5 = 0

for i in range (len(list_X3)):
  temp = list_X3[i] * list_X5[i]
  jum_X3xX5 += temp

print(jum_X3xX5)

345463.0


In [ ]:
jum_X4xX5 = 0

for i in range (len(list_X4)):
  temp = list_X4[i] * list_X5[i]
  jum_X4xX5 += temp

print(jum_X4xX5)

1532176.0


In [ ]:
jum_X1xY = 0

for i in range (len(list_X1)):
  temp = list_X1[i] * list_Y[i]
  jum_X1xY += temp

print(jum_X1xY)

2.21080188110875e+25


In [ ]:
jum_X2xY = 0

for i in range (len(list_X2)):
  temp = list_X2[i] * list_Y[i]
  jum_X2xY += temp

print(jum_X2xY)

2.8900609231781503e+25


In [ ]:
jum_X3xY = 0

for i in range (len(list_X3)):
  temp = list_X3[i] * list_Y[i]
  jum_X3xY += temp

print(jum_X3xY)

3335102462000000.0


In [ ]:
jum_X4xY = 0

for i in range (len(list_X4)):
  temp = list_X4[i] * list_Y[i]
  jum_X4xY += temp

print(jum_X4xY)

8430132552000000.0


In [ ]:
jum_X5xY = 0

for i in range (len(list_X5)):
  temp = list_X5[i] * list_Y[i]
  jum_X5xY += temp

print(jum_X5xY)

4076819971000000.0


## Membuat Matriks A untuk mencari SPL




In [ ]:
A = np.array([[len(df), jum_X1,jum_X2,jum_X3,jum_X4,jum_X5],
    [jum_X1,jum_X1pangkat2,jum_X1xX2,jum_X1xX3,jum_X1xX4,jum_X1xX5],
    [jum_X2,jum_X1xX2,jum_X2pangkat2,jum_X2xX3,jum_X2xX4,jum_X2xX5],
    [jum_X3,jum_X1xX3,jum_X2xX3,jum_X3pangkat2,jum_X3xX4,jum_X3xX5],
    [jum_X4,jum_X1xX4,jum_X2xX4,jum_X3xX4,jum_X4pangkat2,jum_X4xX5],
    [jum_X5,jum_X1xX5,jum_X2xX5,jum_X3xX5,jum_X4xX5,jum_X5pangkat2]
])
A

array([[1.16000000e+02, 8.61824520e+12, 8.61751800e+12, 4.56000000e+03,
        2.15510000e+04, 8.32200000e+03],
       [8.61824520e+12, 3.17586339e+24, 3.42113201e+24, 4.12566554e+14,
        1.45597673e+15, 5.70587119e+14],
       [8.61751800e+12, 3.42113201e+24, 4.22171534e+24, 4.07058438e+14,
        1.45569416e+15, 5.65991590e+14],
       [4.56000000e+03, 4.12566554e+14, 4.07058438e+14, 3.79042000e+05,
        8.56478000e+05, 3.45463000e+05],
       [2.15510000e+04, 1.45597673e+15, 1.45569416e+15, 8.56478000e+05,
        5.07973100e+06, 1.53217600e+06],
       [8.32200000e+03, 5.70587119e+14, 5.65991590e+14, 3.45463000e+05,
        1.53217600e+06, 1.06723400e+06]])

## Koefisien Variabel

In [ ]:
b = np.array([jum_Y, jum_X1xY, jum_X2xY, jum_X3xY, jum_X4xY, jum_X5xY])

x = solve(A, b)

SPL = pd.DataFrame(x, index=['b0', 'b1', 'b2', 'b3', 'b4', 'b5'], columns=['Nilai Koefisien'])
SPL

<ipython-input-37-52ff1039b3dc>:3: LinAlgWarning: Ill-conditioned matrix (rcond=2.08921e-24): result may not be accurate.
  x = solve(A, b)


Nilai Koefisien
b0    -6.061183e+10
b1    -3.550117e+00
b2     9.604364e+00
b3     3.972849e+09
b4    -7.600573e+08
b5     9.022980e+08

## SPL regresi berganda

In [ ]:
b0 = x[0]
b1 = x[1]
b2 = x[2]
b3 = x[3]
b4 = x[4]
b5 = x[5]

print("SPL : Y =", b0, " + " , b1,"X1 + ",b2,"X2 + ",b3,"X3 + ",b4,"X4 + ",b5,"X5")

SPL : Y = -60611829069.70882  +  -3.550117458142596 X1 +  9.604363680230055 X2 +  3972849141.353929 X3 +  -760057342.2227511 X4 +  902297971.6293895 X5


## Koefisien Determinasi Berganda (R^2) untuk Y

Persamaan Regresi dalam Matriks

In [ ]:
# Membuat matriks X dan array GDP
X = np.array([[1, x1, x2, x3, x4, x5] for x1, x2, x3, x4, x5 in zip(df['X1'], df['X2'], df['X3'], df['X4'], df['X5'])])
Y = np.array(df['Y'])

# Menggunakan matriks X invers untuk menghitung nilai koefisien
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)

# Membuat DataFrame untuk menampilkan nilai koefisien
koef_b = pd.DataFrame(b, index=['b0', 'b1', 'b2', 'b3', 'b4', 'b5'], columns=['Nilai koefisien'])
koef_b



Nilai koefisien
b0    -6.061183e+10
b1    -3.550117e+00
b2     9.604364e+00
b3     3.972849e+09
b4    -7.600573e+08
b5     9.022980e+08

menghitung JKT

In [ ]:
JKT = 0
JKT = (jum_Ypangkat2) - (jum_Y)**2 / len(Y)
JKT

2.1754422190583974e+26

Menghitung nilai g/galat

In [ ]:
g = X.T.dot(Y)
g

array([5.43841420e+13, 2.21080188e+25, 2.89006092e+25, 3.33510246e+15,
       8.43013255e+15, 4.07681997e+15])

menghitung JKR

In [ ]:
JKR = 0

for i in range(len(b)):
  JKR += b[i] * g[i]
JKR = JKR - (jum_Y)**2 / len(Y)

JKR



1.8081370438252327e+26

Menghitung JKG

In [ ]:
JKG = JKT - JKR
JKG

3.6730517523316474e+25

memperoleh nilai R(Determinasi berganda) dari Y(target)

In [ ]:
R2y1 = JKR / JKT
R2y1

0.8311583860902789

# **Metode Regresi Bertahap**

### ***Step-1***

menghitung nilai R dari setiap X(variabel bebas), lalu dicari mana yang terbesar

In [ ]:
# Menghitung nilai Rmax
Rmax = []
for var in ['X1', 'X2', 'X3', 'X4', 'X5']:
    X = np.array([[1, df[var][i]] for i in range(len(df))])
    Y = np.array(df['Y'])
    b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
    g = X.T.dot(Y)


    JKR = 0
    for i in range(len(b)):
        JKR += b[i] * g[i]
    JKR = JKR - (jum_Y**2) / len(Y)

    Rmax.append(JKR)

# Membuat DataFrame untuk menampilkan nilai Rmax
R = pd.DataFrame(Rmax, index=['X1', 'X2', 'X3', 'X4', 'X5'], columns=['nilai R'])
R






nilai R
X1  1.287427e+26
X2  1.725640e+26
X3  7.174602e+24
X4  2.603359e+24
X5  6.530034e+22

In [ ]:
# Mengambil nilai R terbesar
R[R['nilai R'] == R['nilai R'].max()]

nilai R
X2  1.725640e+26

In [ ]:
# Menghitung nilai variansi residu (s^2) pada X2
X = np.array([[1, df['X2'][i]] for i in range(len(df))])
Y = np.array(df['Y'])
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
g = X.T.dot(Y)

JKT = jum_Ypangkat2 - jum_Y**2 / len(Y)
JKR = np.sum(b * g) - jum_Y**2 / len(Y)
JKG = JKT - JKR
s2 = JKG / (len(Y) - 2)

print ("s^2(X2) =", s2)

# Menghitung nilai F hitung dan F tabel
Fhitung = JKR / s2
Ftabel = stats.f.isf(0.05, 1, len(Y) - 2)

# Menampilkan hasil
if Fhitung > Ftabel:
    print('f hitung =', Fhitung, '>', 'f tabel =', Ftabel)
    print('Maka peubah X2 masuk dalam model')
else:
    print('f hitung =', Fhitung, '<', 'f tabel =', Ftabel)
    print('Maka peubah X2 tidak masuk dalam model')

s^2(X2) = 3.9456352914563145e+23
f hitung = 437.35410608501843 > f tabel = 3.924330484639666
Maka peubah X2 masuk dalam model


### ***Step-2***

X2 sudah diambil sebagai variabel yang termasuk model, sehingga tersisa X1, X3, X4, X5

In [ ]:
# Menghitung nilai R2max
R2max = []
for var in ['X1', 'X3', 'X4', 'X5']:
    X = np.array([[1, df['X2'][i], df[var][i]] for i in range(len(df))])
    Y = np.array(df['Y'])
    b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
    g = X.T.dot(Y)

    JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
    R2max.append(JKR - Rmax[1])

# Membuat DataFrame untuk menampilkan nilai R2max
R2 = pd.DataFrame(R2max, index=['X1', 'X3', 'X4', 'X5'], columns=['nilai R'])
R2


nilai R
X1  4.055902e+24
X3  2.634671e+24
X4  4.132691e+23
X5  6.162228e+23

In [ ]:
# R terbesar yang kedua
R2[R2['nilai R'] == R2['nilai R'].max()]

nilai R
X1  4.055902e+24

In [ ]:
# Menghitung nilai s^2 dari X1
X = np.array([[1, df['X2'][i], df['X1'][i]] for i in range(len(df))])
Y = np.array(df['Y'])
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
g = X.T.dot(Y)

JKT = jum_Ypangkat2 - (jum_Y**2) / len(Y)
JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
JKG = JKT - JKR
s2 = JKG / (len(Y) - 3)

print("s2(X1) =", s2)

# F hitung dan F tabel
Fhitung = (JKR-Rmax[1]) / s2
Ftabel = stats.f.isf(0.05, 1, len(Y) - 3)

if Fhitung > Ftabel:
  print('f hitung = ', Fhitung,'>', 'f tabel = ', Ftabel)
  print('Maka peubah X1 masuk dalam model')
else:
  print('f hitung = ', Fhitung,'<', 'f tabel = ', Ftabel)
  print('Maka peubah X1 tidak masuk dalam model')


s2(X1) = 3.621623057851165e+23
f hitung =  11.199127308655711 > f tabel =  3.9250756165304113
Maka peubah X1 masuk dalam model


### ***Step-3***

X1 sudah diambil sebagai variabel yang termasuk model, sehingga tersisa X3, X4, X5

In [ ]:
# Menghitung nilai R3max
R3max = []
for var in ['X3', 'X4', 'X5']:
    X = np.array([[1, df['X2'][i], df['X1'][i], df[var][i]] for i in range(len(df))])
    Y = np.array(df['Y'])
    b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
    g = X.T.dot(Y)

    JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
    R3max.append(JKR - (Rmax[1] + R2max[0]))

# Membuat DataFrame untuk menampilkan nilai R3max
R3 = pd.DataFrame(R3max, index=['X3', 'X4', 'X5'], columns=['nilai R'])
R3


nilai R
X3  3.172648e+24
X4  5.571913e+23
X5  5.638811e+23

In [ ]:
R3[R3['nilai R'] == R3['nilai R'].max()]

nilai R
X3  3.172648e+24

In [ ]:
# Menghitung nilai s^2 dari X1.2 dan X3
X = np.array([[1, df['X2'][i], df['X1'][i], df['X3'][i]] for i in range(len(df))])
Y = np.array(df['Y'])
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
g = X.T.dot(Y)

JKT = jum_Ypangkat2 - (jum_Y**2) / len(Y)
JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
JKG = JKT - JKR
s2 = JKG / (len(Y) - 4)

print("s^2(X3) =",s2)

# F hitung dan F tabel
Fhitung = (JKR - (Rmax[1] + R2max[0])) / s2
Ftabel = stats.f.isf(0.05, 1, len(Y) - 4)

if Fhitung > Ftabel:
  print('f hitung = ', Fhitung,'>', 'f tabel = ', Ftabel)
  print('Maka peubah X3 masuk dalam model')
else:
  print('f hitung = ', Fhitung,'<', 'f tabel = ', Ftabel)
  print('Maka peubah X3 tidak masuk dalam model')


s^2(X3) = 3.370686819720839e+23
f hitung =  9.412467970274154 > f tabel =  3.9258342687862307
Maka peubah X3 masuk dalam model


### ***Step-4***

X3 sudah diambil sebagai variabel yang termasuk model, sehingga tersisa X4, X5

In [ ]:
# Menghitung nilai R4max
R4max = []
for var in ['X4', 'X5']:
    X = np.array([[1, df['X2'][i], df['X1'][i], df[var][i], df['X3'][i]] for i in range(len(df))])
    Y = np.array(df['Y'])
    b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
    g = X.T.dot(Y)

    JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
    R4max.append(JKR - (Rmax[1] + R2max[0] + R3max[0]))

# Membuat DataFrame untuk menampilkan nilai R4max
R4 = pd.DataFrame(R4max, index=['X4', 'X5'], columns=['nilai R'])
R4


nilai R
X4  6.409456e+23
X5  4.056220e+23

In [ ]:
R4[R4['nilai R'] == R4['nilai R'].max()]

nilai R
X4  6.409456e+23

In [ ]:
# Menghitung nilai s^2 dari X1.2, X3, X4
X = np.array([[1, df['X2'][i], df['X1'][i], df['X3'][i], df['X4'][i]] for i in range(len(df))])
Y = np.array(df['Y'])
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
g = X.T.dot(Y)

JKT = jum_Ypangkat2 - (jum_Y**2) / len(Y)
JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
JKG = JKT - JKR
s2 = JKG / (len(Y) - 5)

print("s^2(X4) =", s2)

# F hitung dan F tabel
Fhitung = (JKR - (Rmax[1] + R2max[0]+ R3max[0])) / s2
Ftabel = stats.f.isf(0.05, 1, len(Y) - 5)

if Fhitung > Ftabel:
  print('f hitung = ', Fhitung,'>', 'f tabel = ', Ftabel)
  print('Maka peubah X4 masuk dalam model')
else:
  print('f hitung = ', Fhitung,'<', 'f tabel = ', Ftabel)
  print('Maka peubah X4 tidak masuk dalam model')


s^2(X4) = 3.343310517309668e+23
f hitung =  1.9170987421528867 < f tabel =  3.926606812744345
Maka peubah X4 tidak masuk dalam model


In [ ]:
# Menghitung nilai s^2 dari X1, X2, X3, dan X5
X = np.array([[1, df['X1'][i], df['X2'][i], df['X3'][i], df['X5'][i]] for i in range(len(df))])
Y = np.array(df['Y'])
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
g = X.T.dot(Y)

JKT = jum_Ypangkat2 - (jum_Y**2) / len(Y)
JKR = np.sum(b * g) - (jum_Y**2) / len(Y)
JKG = JKT - JKR
s2 = JKG / (len(Y) - 5)

print("s^2(X5) =", s2)

# F hitung dan F tabel
Fhitung = (JKR - (Rmax[1] + R2max[0]+ R3max[0])) / s2
Ftabel = stats.f.isf(0.05, 1, len(Y) - 5)

if Fhitung > Ftabel:
  print('f hitung = ', Fhitung,'>', 'f tabel = ', Ftabel)
  print('Maka peubah X5 masuk dalam model')
else:
  print('f hitung = ', Fhitung,'<', 'f tabel = ', Ftabel)
  print('Maka peubah X5 tidak masuk dalam model')


s^2(X5) = 3.364510845395284e+23
f hitung =  1.2055898037626755 < f tabel =  3.926606812744345
Maka peubah X5 tidak masuk dalam model


In [ ]:
# Mencari nilai b
X = np.array([[1, df['X2'][i], df['X1'][i], df['X3'][i]] for i in range(len(df))])
Y = np.array(df['Y'])
b = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)

# Konversi ke dataframe
df_b = pd.DataFrame(b, index=['b0', 'b1', 'b2', 'b3'], columns=['Nilai Koefisien'])
df_b



Nilai Koefisien
b0    -1.405600e+11
b1     9.585488e+00
b2    -3.503946e+00
b3     4.009617e+09

### ***SPL regresi bertahap***

In [ ]:
b0 = b[0]
b1 = b[1]
b2 = b[2]
b3 = b[3]

print("SPL : Y =", b0, " + " , b1,"X1 + ",b2,"X2 + ",b3,"X3")

SPL : Y = -140560004547.3942  +  9.585487627972567 X1 +  -3.5039457618365573 X2 +  4009617126.683123 X3


### ***Koefisien Determinasi berganda(R^2)***

In [ ]:

# Menghitung JKT
JKT = jum_Ypangkat2 - (jum_Y**2) / len(Y)

# Menghitung JKR
JKR = 0
for i in range(len(df_b)):
    JKR += b[i] * g[i]  # g = X.T.dot(Y)
JKR = JKR - (jum_Y**2) / len(Y)

# Menghitung R2
R2y2 = JKR / JKT
R2y2




0.41776043910912597

# **Kesimpulan**

Dalam program ini kami membuat analisis regresi berganda menggunakan metode regresi bertahap, bersumber dari dataset factbook.csv. hasil analisis yang kami dapat diantaranya :

1. SPL : Y = -140560004547.3942  +  9.585487627972567 X1 +  -3.5039457618365573 X2 +  4009617126.683123 X3 (regresi bertahap)
2. koefisien determinasi untuk metode regresi bertahap = 0.41776043910912597
3. Variabel yang memiliki hubungan yang signifikan : X1(Exports), X2 (Imports), X3(Industrial production growth rate).




